lib/iris/plot.py

```diff
@@ -52,9 +54,28 @@
 # Cynthia Brewer citation text.
 BREWER_CITE = 'Colours based on ColorBrewer.org'
 
-
 PlotDefn = collections.namedtuple('PlotDefn', ('coords', 'transpose'))
 
+# Threading reentrant lock to ensure thread-safe plotting.
+_lock = threading.RLock()
+
+
+def _locker(func):
+    """
+    Decorator that ensures a thread-safe atomic operation is
+    performed by the decorated function.
+
+    Uses a shared threading reentrant lock to provide thread-safe
+    plotting by public API functions.
+
+    """
+    @wraps(func)
+    def decorated_func(*args, **kwargs):
+        with _lock:
+            result = func(*args, **kwargs)
+        return result
+    return decorated_func
+
```

Comment:
A simple decorator to facilitate the use of a re-entrant lock to provide thread-safe plotting. A re-entrant lock is required as `iris.plot.contourf` calls `iris.plot.contour` ...